### DISCUSSION TITLE
(k3s) Finding a starting point...

### DISCUSSION THREAD

**Discussion by IngwiePhoenix** [concourse/concourse#8959] [General]
Hello there!

First, for a little bit of context: I started my journey into Kubernetes late november of last year and I have now finally gotten around to deploy my first node with k3s; in total, this will span three arm64 and one riscv64 node. However, some images that I want to use don't have a riscv64 version - and some other applications I wish to deploy don't even have their own CI/CD for container images at all.

So I need to make my own! And after crawling ArtifactHub and many other k8s related resources, I ended up here.

I read through the documentation and I believe to have (finally!) found what I was looking for: A way to specify a remote, not owned by me, Git repository and then run arbitrary commands, in order, to do something with it.

But I am a little bit lost on a few details that I either didn't really get from the docs or that I might have misunderstood. Hence why I found my way here, hoping to find a few pointers. :)

There are two main pipelines that I want to run for starting out:

* Snort (Nostr client)
  - Clone https://git.v0l.io/Kieran/snort (master, on each commit - it's basically rolling)
  - Download my self-crafted Dockerfile (not checked into a repo yet) and place it into the repo
  - Use [Kaniko](https://github.com/GoogleContainerTools/kaniko?tab=readme-ov-file) to process the build and publish
* k3s itself, for riscv64
  - Clone the k3s repository pinned to a (the latest) tag
  - Apply the mini-patches from [my instructions in the related ticket](https://github.com/k3s-io/k3s/issues/7151#issuecomment-1978539620) possibly by using a shell script in an Alpine container
  - Execute the build
  - Collect the resulting `k3s` binary and publish it... somewhere?

Both of them are straight forward; just resources pointing to Git repositories to trigger a set of jobs when ready.

Here is what I haven't managed to pick up from the docks yet:

- How would I implement an entire shell script as a step? Patching k3s is simple, but I do need to run a few commands. Normally, I would drop it into a `ConfigMap` and mount that.
- So far, I had only see how to reference Git repos by their branch -. but not specifically by their tag.
- How can I add node selectors to primarily run my builds on the arm64 nodes (`matchLabels: { kubernetes.io/arch: arm64 }`) or on the riscv64 node respectively? As far as I understand, jobs are just a series of starting and stopping pods - so they are subjected to the regular scheduler.
- Does Concourse have a built-in artifact collection or webUI from where people could possibly download that from, or do I have to host my own?

Thank you a lot and kind regards,

Ingwie

**Comment by taylorsilva** [comment]
Hey, sorry for the late reply.

So one issue I'm seeing right off the bat, assuming you want to run Concourse on your k3s cluster, is that your k3's cluster is running on arm64 machines. Sadly Concourse doesn't officially release arm64 binaries at this time. I know some community folks build and publish some binaries though.

Now for your questions:

> How would I implement an entire shell script as a step? Patching k3s is simple, but I do need to run a few commands. Normally, I would drop it into a ConfigMap and mount that.

So how this typically works is you make a job that downloads a git repo that contains your shell script, and then execute the shell script in a `task` step. You can see how we do that here: https://github.com/concourse/ci

More concretely, you can see here:
https://github.com/concourse/ci/blob/34456877fca8c3a65d3b1d1e24418edab7c853b3/pipelines/release.yml#L142-L145
We `get` the `ci` repo, and then execute the task config at `tasks/yarn-test.yml`, which is executing the script at `tasks/scripts/yarn-test` in the `ci` repo. That's generally how you execute a script as a step within Concourse.

> So far, I had only see how to reference Git repos by their branch -. but not specifically by their tag.

You can see other ways to configure resources. In this case you want to read the docs for the git resource: https://github.com/concourse/git-resource/

You'll see that it has `tag_filter` and `tag_regex` fields that you can set in place of the `branch` param.

> How can I add node selectors to primarily run my builds on the arm64 nodes (matchLabels: { kubernetes.io/arch: arm64 }) or on the riscv64 node respectively? As far as I understand, jobs are just a series of starting and stopping pods - so they are subjected to the regular scheduler.

Concourse doesn't use the k8s scheduler when you run Concourse on k8s. Concourse schedules and manages its own containers. The guidance for running Concourse workers on k8s is that each worker has the entire capacity of a single k8s node. You can ignore this guidance if you're _very careful_ and ensure you don't run anything that's too cpu/memory intensive. This is a bad idea in general though.

> Does Concourse have a built-in artifact collection or webUI from where people could possibly download that from, or do I have to host my own?

Concourse does not have built-in artifact collection. You need to push your artifacts somewhere before your job finishes, otherwise they will be deleted from disk shortly after your job finishes.

Hope that helps. Feel free to ask more questions.

**Reply by IngwiePhoenix** [reply]
> Sadly Concourse doesn't officially release arm64 binaries at this time.

Oh that is very unfortunate... Yes, my cluster is indeed arm64 and will soon be joined by riscv64 as well. So I will have to look into either building my own images or see what the community has made so far. Oof... more custom builds, yay. x)

> (...) and then execute the shell script in a task step.

I took a look and although this is quite a heavy hitter for an example, I think I figured out what you ment. I just haven't fully found out how the individual files in `task/` get loaded; like, where is the root of the whole operation? That said, that's quite nice. I could package custom dockerfiles and scripts in there, and just reference a task. This does pretty much all I would want in terms of custom builds without affecting the upstream repository. Thanks!

> Concourse schedules and manages its own containers.

So it does not actually spawn new pods and runs everything in the same context? That's interesting. I would have thought that even just running a simple `kubectl run -n $NS --image $IMAGE (...)` would be enough to spawn a new pod and thus use the scheduler. Alright, then I will have to make sure to not run intense actions in parallel (like building a Linux kernel...which I do, sometimes, for riscv64). Thanks for the info!

> You need to push your artifacts somewhere before your job finishes

Fair enough, there's a plethora of tools like Sharry, Zipline or even more specialized ones out there. :) Hey, it could even be a resource I can `put` to, right?


Thank you for the detailed answer! I have gotten Tekton to install in the meantime, but it is much more complicated and I like the Concourse approach more in terms of layout.

I will look into the images and see what I can put together. If I come across anything, I will let you know!

Thanks and kind regards,
Ingwie
